File: android/public-app/src/main/kotlin/hu/orszembejelento/app/report/data/CatalogRepository.kt
package hu.orszembejelento.app.report.data

import hu.orszembejelento.app.report.data.network.PublicApi

data class CatalogEventType(val code: String, val displayName: String)
data class CatalogCategory(val code: String, val displayName: String, val eventTypes: List<CatalogEventType>)

sealed class CatalogResult {
    data class Loaded(val categories: List<CatalogCategory>) : CatalogResult()
    data object Failed : CatalogResult()
}

/**
 * The Public event catalogue - always fetched from the backend, never hardcoded (§3, §44).
 *
 * Caches the result in memory for the current process only (§45 - no durable offline
 * catalogue is required), so re-entering the report flow within one app session does not
 * refetch on every recomposition.
 */
class CatalogRepository(private val api: PublicApi) {

    @Volatile
    private var cached: List<CatalogCategory>? = null

    suspend fun catalog(forceRefresh: Boolean = false): CatalogResult {
        if (!forceRefresh) {
            cached?.let { return CatalogResult.Loaded(it) }
        }

        val response = runCatching { api.reportCatalog() }.getOrNull()
        val body = response?.takeIf { it.isSuccessful }?.body() ?: return CatalogResult.Failed

        val categories = body.categories.map { category ->
            CatalogCategory(
                code = category.code,
                displayName = category.displayName,
                eventTypes = category.eventTypes.map { CatalogEventType(it.code, it.displayName) },
            )
        }
        cached = categories
        return CatalogResult.Loaded(categories)
    }
}

File: web/public-web/src/repository/catalogRepository.ts
import { publicApi, type PublicApi } from '../api/publicApi'

export interface CatalogEventType {
  readonly code: string
  readonly displayName: string
}
export interface CatalogCategory {
  readonly code: string
  readonly displayName: string
  readonly eventTypes: readonly CatalogEventType[]
}

export type CatalogResult = { readonly kind: 'loaded'; readonly categories: readonly CatalogCategory[] } | { readonly kind: 'failed' }

/**
 * The Public event catalogue - always fetched from the backend, never hardcoded (§3, §44).
 * Caches the result in memory for the current page session only (§45 - no durable offline
 * catalogue is required).
 */
export function createCatalogRepository(api: PublicApi = publicApi) {
  let cached: readonly CatalogCategory[] | null = null

  async function catalog(forceRefresh = false): Promise<CatalogResult> {
    if (!forceRefresh && cached) return { kind: 'loaded', categories: cached }

    const result = await api.reportCatalog().catch(() => undefined)
    if (!result || result.status !== 200 || !result.body) return { kind: 'failed' }

    const categories = result.body.categories.map((category) => ({
      code: category.code,
      displayName: category.displayName,
      eventTypes: category.eventTypes.map((eventType) => ({ code: eventType.code, displayName: eventType.displayName })),
    }))
    cached = categories
    return { kind: 'loaded', categories }
  }

  return { catalog }
}

export type CatalogRepository = ReturnType<typeof createCatalogRepository>
